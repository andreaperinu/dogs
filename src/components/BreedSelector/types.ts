import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
