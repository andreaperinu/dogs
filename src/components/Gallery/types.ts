import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
