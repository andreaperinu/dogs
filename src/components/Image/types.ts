import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
