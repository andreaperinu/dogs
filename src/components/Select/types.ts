import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
