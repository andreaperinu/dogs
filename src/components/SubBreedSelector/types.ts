import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
