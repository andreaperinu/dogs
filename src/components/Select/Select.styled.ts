import styled, { css } from 'styled-components';

import { ISelectStyledProps } from './types';

export const SelectStyled = styled.div<ISelectStyledProps>`
${({ showOptions, disabled, theme: {
	border, borderRadius, minHeight, brandPrimary
} }) => css`

		position: relative;
		display: flex;
		align-items: center;

		.wrapper {
			width: 100%;
			max-width: 20rem;
			position: relative;
		}

		.list {
			opacity: 0;
			height: 0;
      margin: 0;
			padding: 0;
			transition: opacity .5s;
			overflow: hidden;
			list-style-type: none;
			background: white;
    	max-height: 20rem;
    	overflow: auto;
			
			${showOptions && css`
				border: ${border};
				height: auto;
				opacity: 1;
			`}
		}

		.value {
			border: ${border};
			border-radius: ${borderRadius};
      padding: .5rem;
			min-height: ${minHeight};
			box-sizing: border-box;
			white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;

			${!disabled && css` cursor: pointer; `}
		}

		li {
      padding: .5rem;

			${showOptions && css` cursor: pointer; `}
			&:hover { background: ${brandPrimary}; }
		}

		.list {
			border-top: none;
			position: absolute;
			width: 100%;
			box-sizing: border-box;
			z-index: 1;
		}

		.text {
			${!disabled && css` cursor: pointer; `}
			margin-left: .5rem;
		}

	`}
`
