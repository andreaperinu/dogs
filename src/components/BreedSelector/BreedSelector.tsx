import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { useCallback } from 'react';

import {
	maybeBreedBySelectItem, maybeBreedsToSelectItems, maybeBreedToSelectItem
} from '@/utils/utils';

import { Select } from '../Select/Select';
import { ISelectItem } from '../Select/types';
import { IBreedSelectorProps } from './types';

export const BreedSelector = ({
	maybeSelectedBreed, maybeBreeds, onChange,
}: IBreedSelectorProps) => {
	
	const onChangeHandler = useCallback((maybeSelectItem: O.Option<ISelectItem>) => pipe(
		maybeSelectItem,
		maybeBreedBySelectItem(maybeBreeds),
		onChange,
	), [maybeBreeds])

	return (
		<>
			<h2>Select a breed</h2>
			<Select
				maybeValue={pipe(maybeSelectedBreed, maybeBreedToSelectItem)}
				maybeOptions={pipe(maybeBreeds, maybeBreedsToSelectItems)}
				onChange={onChangeHandler}
				data-test="component-select"
			/>
		</>
	)
}
