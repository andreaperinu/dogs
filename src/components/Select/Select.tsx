import classNames from 'classnames';
import * as A from 'fp-ts/lib/Array';
import * as B from 'fp-ts/lib/boolean';
import { constVoid, pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React, { memo, useCallback, useMemo } from 'react';

import { noopComponent } from '@/utils/utils';

import { SelectStyled } from './Select.styled';
import { ISelectItem, ISelectProps } from './types';

export const BaseSelect = ({
  maybeValue, maybeOptions, className, onChange, disabled = false, maybeText = O.none,
}: ISelectProps) => {
  const [showOptions, setShowOptions] = React.useState(false)

  const onItemSelect = useCallback((item: ISelectItem) => {
    setShowOptions(false)
    onChange(O.fromNullable(item))
  }, [onChange])

  const onShowOptionsHandler = useCallback(() => pipe(
    !disabled,
    B.fold(constVoid, () => setShowOptions((prev) => !prev))
  ), [disabled])

  const renderedValue = useMemo(() => pipe(
    maybeValue,
    O.map(({ value }) => value),
    O.getOrElse(() => ''),
    v => <div className={classNames('value', className)} onClick={onShowOptionsHandler}>{v}</div>
  ), [maybeValue, onShowOptionsHandler])

  const renderedOptions = useMemo(() => pipe(
    maybeOptions,
    O.map(A.map(option => (
      <li key={option.id} value={option.id} onClick={() => onItemSelect(option)}>
        {option.value}
      </li>
    ))),
    O.map(O.fromPredicate(A.isNonEmpty)),
    O.chain(O.map(items => <ul className="list">{items}</ul>)),
    O.getOrElse(noopComponent),
  ), [maybeOptions, onItemSelect])

  const renderedText = useMemo(() => pipe(
    maybeText,
    O.map(text => <div className="text" onClick={onShowOptionsHandler}>{text}</div>),
    O.getOrElse(noopComponent)
  ), [maybeText, onShowOptionsHandler])

  return (
    <SelectStyled showOptions={showOptions} disabled={disabled}>

      <div className="wrapper">
        {renderedValue}
        {renderedOptions}
      </div>

      {renderedText}
    </SelectStyled>
  )
}

export const Select = memo(BaseSelect)
