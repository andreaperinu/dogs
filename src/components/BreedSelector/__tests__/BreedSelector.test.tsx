import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React from 'react';

import { ISelectItem } from '@/components/Select/types';
import { IBreed } from '@/models/models';
import * as mockedBreeds from '@/tests/mockedBreeds.json';
import { findByDataTest, setup } from '@/tests/utils';
import { formatBreeds, maybeBreedToSelectItem } from '@/utils/utils';

import { BreedSelector as component } from '../BreedSelector';

const onChange = jest.fn()

const maybeBreeds: O.Option<IBreed[]> = pipe(
	mockedBreeds,
	formatBreeds,
	O.fromNullable,
)

const maybeSelectedBreed: O.Option<IBreed> = pipe(maybeBreeds, O.chain(A.head))

const props: React.ComponentProps<typeof component> = {
	maybeBreeds,
	maybeSelectedBreed,
	onChange,
}

describe('<BreedSelector />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props })
		expect(wrapper).toMatchSnapshot()
	})

	it('should set breed', () => {
		const mockedMaybeSelectItem: O.Option<ISelectItem> = pipe(
			maybeSelectedBreed,
			maybeBreedToSelectItem,
		)

		const wrapper = setup({ component, props })
		const select = findByDataTest(wrapper, 'select')

		select?.simulate('change', mockedMaybeSelectItem)

		expect(onChange).toHaveBeenCalledWith(maybeSelectedBreed)
	})
})