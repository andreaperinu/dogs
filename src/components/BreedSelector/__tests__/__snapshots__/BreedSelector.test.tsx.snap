// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreedSelector /> should match snapshot 1`] = `
<Fragment>
  <h2>
    Select a breed
  </h2>
  <Memo(BaseSelect)
    data-test="component-select"
    maybeOptions={
      Object {
        "_tag": "Some",
        "value": Array [
          Object {
            "id": "affenpinscher",
            "value": "affenpinscher",
          },
          Object {
            "id": "african",
            "value": "african",
          },
          Object {
            "id": "airedale",
            "value": "airedale",
          },
          Object {
            "id": "akita",
            "value": "akita",
          },
          Object {
            "id": "appenzeller",
            "value": "appenzeller",
          },
          Object {
            "id": "australian",
            "value": "australian",
          },
          Object {
            "id": "bulldog",
            "value": "bulldog",
          },
          Object {
            "id": "terrier",
            "value": "terrier",
          },
          Object {
            "id": "default",
            "value": "default",
          },
        ],
      }
    }
    maybeValue={
      Object {
        "_tag": "Some",
        "value": Object {
          "id": "affenpinscher",
          "value": "affenpinscher",
        },
      }
    }
    onChange={[Function]}
  />
</Fragment>
`;
